cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,313 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#include "UrdfRobot/UrdfRobotComponent.h"
+
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <Containers/Array.h>       // TArray
+#include <UObject/Object.h>         // CreateDefaultSubobject
+#include <UObject/UObjectGlobals.h> // NewObject
+
+#include "CoreUtils/ArrayDesc.h"
+#include "CoreUtils/Assert.h"
+#include "CoreUtils/Config.h"
+#include "CoreUtils/InputActionComponent.h"
+#include "CoreUtils/Log.h"
+#include "CoreUtils/Std.h"
+#include "CoreUtils/Unreal.h"
+#include "UrdfRobot/UrdfJointComponent.h"
+#include "UrdfRobot/UrdfLinkComponent.h"
+#include "UrdfRobot/UrdfParser.h"
+
+// useful for debugging horizontal_pendulum.urdf
+const std::map<std::string, std::map<std::string, std::vector<double>>> DEFAULT_INPUT_ACTIONS = {
+    {"One",   {{"joint_0.add_torque_in_radians",          { 1000000.0, 0.0, 0.0}}}},
+    {"Two",   {{"joint_0.add_torque_in_radians",          {-1000000.0, 0.0, 0.0}}}},
+    {"Three", {{"joint_1.add_to_angular_velocity_target", { 0.1,       0.0, 0.0}}}},
+    {"Four",  {{"joint_1.add_to_angular_velocity_target", {-0.1,       0.0, 0.0}}}}
+};
+
+UUrdfRobotComponent::UUrdfRobotComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    // In the current setup, the UrdfRobotComponent and the UrdfBotPawn (UrdfRobotComponent is the RootComponent of this Pawn) 
+    // do not move along with the movement of child LinkComponents. However, we want the UrdfBotPawn's pose to update/follow 
+    // the pose of child LinkComponents. Hence, we enable this component to tick and in every tick we update it's pose to 
+    // follow the root link's pose.
+    PrimaryComponentTick.bCanEverTick = true;
+    PrimaryComponentTick.TickGroup = ETickingGroup::TG_PostPhysics;
+
+    // UInputActionComponent
+    input_action_component_ = CreateDefaultSubobject<UInputActionComponent>(Unreal::toFName("input_action_component"));
+    SP_ASSERT(input_action_component_);
+    // Need to explicitly set this up so that the component hierarchy is well-defined.
+    input_action_component_->SetupAttachment(this);
```

Comment:
Do we need to call `RegisterComponent` here?

Reply:
Ah ok. So what's the rule? Always call `SetupAttachment(...)` immediately, and then call `RegisterComponent()` immediately afterwards if the component wasn't created with `CreateDefaultSubobject(...)`?

Reply:
Yes, I think we should make this a rule. If a component is created using `NewObject`, make sure to call `SetupAttachment()` and `RegisterComponent()` immediately after creating it. To my knowledge, this is not necessary when child components are not involved, but, there is no known drawback to this rule, so we might as well have it.

Reply:
I like that rule. Even if this rule only matters when child components are involved, we should make it a rule everywhere. Otherwise I would need to need to look through the entire class to figure out if there is a bug, because a child component might be created anywhere. In contrast, if we enforce this rule everywhere, then I only need to look at 3 lines of code (`a = NewObject(...)`, `SP_ASSERT(a)`, `a->SetupAttachment(...)`) to figure out of there is a bug.